Bump workflow (py-morphology-repair-workflow 2.0.3)

GianlucaFicarelli opened the pull request:
Replace https://github.com/BlueBrain/spack/pull/1058

GianlucaFicarelli committed:
Bump workflow

Bump py-morphology-repair-workflow and update related packages.

asanin-epfl commented on var/spack/repos/builtin/packages/py-morph-tool/package.py:
```diff
@@ -14,24 +14,20 @@ class PyMorphTool(PythonPackage):
+    depends_on('py-deprecation@2.1.0:', type='run')
+    depends_on('py-more-itertools@8.6.0:', type='run')
+    depends_on('py-morphio@2.7.0:', type='run')
+    depends_on('py-neurom@1.8.0:', type='run')
```
You mentioned that would be good to constrain it by upper bound, no? `py-neurom@1.8.0:1.999`?

GianlucaFicarelli replied:
Right, but the CI is still failing and I didn't want to add more reasons to fail, so I'll need to add the constraints later.
However if I'm not wrong `py-morphology-repair-workflow 2.0.3` requires `neurom 2`, so all the other packages required by `py-morphology-repair-workflow` should be compatible with `neurom 2`.
If this is not the case, we should fix them before merging this PR.

asanin-epfl replied:
ok

asanin-epfl commented on var/spack/repos/builtin/packages/py-morph-tool/package.py:
```diff
@@ -14,24 +14,20 @@ class PyMorphTool(PythonPackage):
+    depends_on('py-xmltodict@0.12:', type='run')
+
+    depends_on('py-plotly@4.1:', type='run')
+    depends_on('py-dask+bag@2.19:', type='run')
```
Might the upper constraint bite us as dask switches to calendar versioning since 2.30?

GianlucaFicarelli replied:
Yes, I think it's a good (safer) idea. We should update also `setup.py` in the package repository, though (at the moment it's `dask[bag]>=2.19.0` in the current master at https://github.com/BlueBrain/morph-tool/blob/1dc791e34dc49e3afaf3b689b277b4f6ea774e44/setup.py#L26).

asanin-epfl replied:
ok, anyway it would be better if PR is approved by someone from the spack team. All good from my side.

mgeplf commented:
LGTM; looks like another mpi variant issue blocking this?

GianlucaFicarelli commented:
> LGTM; looks like another mpi variant issue blocking this?

I'm holding this PR waiting for https://github.com/BlueBrain/spack/pull/1100, that should fix this concretization error.
As mentioned in the previous comments I was thinking to pin `neurom<2` in the spack packages requiring `neurom` but not required by `py-morphology-repair-workflow` (i.e. excluding `py-morph-tool` that needs to support also `neurom>=2`).

GianlucaFicarelli commented:
The CI is passing now, but I moved the PR to draft because the following packages are required by `py-morphology-repair-workflow` but they probably need to be updated to be compatible with `neurom >= 2` (i.e. the `neurom` version cannot be pinned to `< 2`):
```
py-bluepy
py-bluepysnap
py-morph-tool
py-morph-validator
```
These packages instead are not required and they could be pinned:
```
py-diameter-synthesis
py-tns
```
Alternatively, I could check if it's possible to make a transitory release in spack (not `2.x`, but something like `wf.0.1`, `wf.0.2` or any better non-numeric naming) to be used only with `py-morphology-repair-workflow` for the following packages (however, it's possible that the packages will not work correctly if they use any of the features changed or removed in `neruom`):
```
py-neurom
py-bluepy
py-bluepysnap
py-morph-tool
py-morph-validator
```

matz-e commented on var/spack/repos/builtin/packages/py-morphology-repair-workflow/package.py:
```diff
@@ -19,8 +19,20 @@ class PyMorphologyRepairWorkflow(PythonPackage):
-    depends_on('py-more-itertools', type='run')
-    depends_on('py-morph-validator', type='run')
+    depends_on('py-click@7.0:', type='run')
+    depends_on('py-dask+bag@2.19.0:', type='run')
```
Is this actually distributed or could this be done via multiprocessing?

GianlucaFicarelli replied:
From what I can see `dask` in `morphology-repair-workflow` is used only in 
```
ok, failures = zip(*dask.bag.from_sequence(iterable).map(wrapped).compute())
```
There is no multi-node setup, but It should wrap multiprocessing.
Do you see any reason to avoid dask in this case? (In some other cases we used `joblib`)

matz-e replied:
IMO, the question should more be why Dask and not just pure `multiprocessing`? Why add a complex dependency when `multiprocessing` could do it?

Same for `joblib`, if it's just to run stuff on a single node.

arnaudon replied:
+1 for multiprocessing if it is not distributed. I have a PR open on that code to fix dependencies with latest versions of other codes, I can remove dask dependency in the same time.

GianlucaFicarelli commented:
Needs https://bbpteam.epfl.ch/project/issues/browse/NSETM-1442 before merging.

GianlucaFicarelli commented:
Retest this please

GianlucaFicarelli commented:
There are errors in Software Tests / Neurodamus manual building at  https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/hpc.module-testing/detail/hpc.module-testing/3480/pipeline

@matz-e do you know if it can be something not related to this PR?

GianlucaFicarelli commented:
> There are errors in Software Tests / Neurodamus manual building at https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/hpc.module-testing/detail/hpc.module-testing/3480/pipeline
> 
> @matz-e do you know if it can be something not related to this PR?

@matz-e or @pramodk , after comparing the log with another successful execution, I see that jenkins for this PR is using an old version of neurodamus-neocortex:
/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/modules/applications/2021-03/tcl/linux-rhel7-x86_64/neurodamus-neocortex/1.3-3.2.0 
instead of
/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/modules/applications/2021-04/tcl/linux-rhel7-x86_64/neurodamus-neocortex/1.4-3.2.2

I think it could be fixed just cleaning the workspace /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/pulls/1099
If this is the case, what is the right way to delete it?

matz-e commented:
> I think it could be fixed just cleaning the workspace /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/pulls/1099
> If this is the case, what is the right way to delete it?

Secret stuff by us :)

matz-e commented:
How about a rebase to retest this? Working directory should be clean now.

GianlucaFicarelli commented:
I rebased and pushed, but there is a new error at https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/hpc.module-testing/detail/hpc.module-testing/3485/pipeline:
```
/jenkins/10/workspace/hpc.module-testing@tmp/durable-a6834299/script.sh:
line 5: /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/pulls/1099/config/modules.sh: No such file or directory
```
Effectively this file is missing, how it should be generated?

pramodk commented:
> Effectively this file is missing, how it should be generated?

thats typically misleading error. If you check & click on the `Test` tab then you will see that there are build errors (and hence module.sh is not generated):

![image](https://user-images.githubusercontent.com/666852/117512161-5a979600-af8f-11eb-9b8a-768bdd3b9cab.png)

GianlucaFicarelli committed:
Bump py-seaborn and py-tqdm

GianlucaFicarelli committed:
Set py-setuptools-scm+toml by default

GianlucaFicarelli committed:
Revert to url instead of pypi

GianlucaFicarelli committed:
Bump py-neurom

GianlucaFicarelli committed:
Fix morphio for lexertl submodule no longer accessible

matz-e commented:
> thats typically misleading error

We could potentially handle this better, but every time I look into issues like this, I'm limited by our outdated Jenkins.

matz-e commented on var/spack/repos/builtin/packages/brainbuilder/package.py:
```diff
@@ -13,39 +13,22 @@ class Brainbuilder(PythonPackage):
     depends_on('py-click@7.0:7.999', type='run')
-    depends_on('py-future@0.16:', type='run')
-    depends_on('py-h5py@2.6:', type='run')
+    depends_on('py-h5py@3.1.0:', type='run')
```
Wow, you just keep bumping the dependencies ;)

matz-e commented on var/spack/repos/builtin/packages/py-bluepy/package.py:
```diff
@@ -40,8 +40,10 @@ class PyBluepy(PythonPackage):
-    depends_on('py-morphio@3.0.0:', type='run', when='@2.3.0:')
+    depends_on('py-morph-tool@2.4.3:2.999', type='run', when='@2.3.0:')
+
+    # morphio
```
Could we in the future leave the obvious comments out?

GianlucaFicarelli replied:
Sure, I think that the bluepy spec will be cleaned when the old versions will be removed.

tomdele replied:
indeed. Not sure if a package still use version of bluepy<2.0.0.

matz-e commented on var/spack/repos/builtin/packages/py-minis-validation/package.py:
```diff
@@ -13,23 +13,22 @@ class PyMinisValidation(PythonPackage):
-    depends_on('py-numpy@1.14:', type='run')
-    depends_on('py-pandas@0.25:', type='run')
+    depends_on('py-numpy@1.14:1.999', type='run')
+    depends_on('py-pandas@0.25:1.999', type='run')
```
Is Pandas 2 going to break things?

GianlucaFicarelli replied:
I'm not sure, because I didn't work on the package, but I enforced the same constraints indicated in `setup.py` for consistency. Maybe it can be discussed outside of this PR and @asanin-epfl can add more comments about that.

asanin-epfl replied:
I've added upper constraints for additional safety. The reason for that was the Dask change to calendar versioning.

GianlucaFicarelli replied:
In this case the constraint is for pandas and numpy, not dask (that's the one usually discussed 😄 ).

asanin-epfl replied:
yes, in order to be consistent among version constraints, I've applied this rule to all dependencies. I'm trying to follow this rule now everywhere as I find it useful to prevent unexpected surprises.

matz-e reviewed (dismissed):
Fine with me.

I remain concerned about all these generic upper limits to enforce the idea of semantic versioning. To me, this seems to assume that something is guaranteed to break, which may not necessarily be the case, but may very well cause us some concretization issues down the road.

GianlucaFicarelli committed:
Other minor bumps

- keep only the latest py-morphio/morphio 3.1.1
- py-neuror from 1.2.2 to 1.2.3
- py-minis-validation from 0.0.4 to 0.0.5

GianlucaFicarelli commented:
Sorry for the new changes, but after re-checking the packages I found that some packages were updated since the previous version. This should be the final update if the build is still successful.

Some notes about morphio:
I removed the old versions of morphio defined in spack because they were built from git, and the repo that was used for the submodule lexertl is no longer accessible, so the build was failing.

This is also causing the main spack build to fail on the default branch, and the latest versions of py-morphio/py-bluepy/py-bluepysnap today are not available in unstable.
A new repo for the submodule has been already set in the master branch of morphio, but a new version has not been released yet.

Changes related to the morphio versions:
- keeping the latest `3.1.1` (because the tar.gz including the submodules is available on pypi, and now it's used during the build)
- removing `3.0.2 2.7.1` (even though they could be kept because the tar.gz is available on pypi. Is there any reason to keep them?
There are several packages depending on morphio, but none of them is pinning morphio < 3 and >= 2.7.1 so they would not be used as dependencies, and py-morphio/2.7.1 can still be loaded from archive/2021-04).
- removing `2.3.10 2.3.4 2.2.1 2.1.2 2.0.8` (because the tar.gz is not available on pypi)

Moreover, I kept `morpho-kit 0.2.0` but I removed `morpho-kit 0.1.0` because it was depending on `morphio<=2.2.1`.

GianlucaFicarelli commented on var/spack/repos/builtin/packages/py-setuptools-scm/package.py:
```diff
@@ -19,7 +19,7 @@ class PySetuptoolsScm(PythonPackage):
     version('1.15.6', sha256='49ab4685589986a42da85706b3311a2f74f1af567d39fee6cb1e088d7a75fb5f')
 
-    variant('toml', default=False, description='Build with TOML support')
+    variant('toml', default=True, description='Build with TOML support')
```
I think it's safe to use the `py-setuptools-scm+toml` by default because it's the preferred way to configure it.
I needed to change it to resolve a concretization issue.

GianlucaFicarelli commented on var/spack/repos/builtin/packages/py-tqdm/package.py:
```diff
@@ -12,12 +12,16 @@ class PyTqdm(PythonPackage):
-    depends_on('py-nose', type='test')
-    depends_on('py-flake8', type='test')
-    depends_on('py-coverage', type='test')
+    variant('telegram', default=False, description='Enable Telegram bot support')
+    variant('notebook', default=False, description='Enable Jupyter Notebook support')
+
+    depends_on('python@2.7:2.8,3.4:', type=('build', 'run'))
+    depends_on('py-setuptools@42:', type=('build', 'run'))
+    depends_on('py-setuptools-scm@3.4:+toml', type='build')
+    depends_on('py-requests', when='+telegram', type=('build', 'run'))
+    depends_on('py-ipywidgets@6:', when='+notebook', type=('build', 'run'))
```
This is mostly from upstream.

asanin-epfl commented on var/spack/repos/builtin/packages/py-neurom/package.py:
```diff
@@ -32,6 +33,10 @@ class PyNeurom(PythonPackage):
 
+    # >= 2.0
+    depends_on('py-morphio@3.1.1:', type='run', when='@2.0:')
+    depends_on('py-pandas@1.0.5:', type='run', when='@2.0:')
```
This line almost repeats the line 42, can it be merged with it?

GianlucaFicarelli replied:
I did in this way because it seemed clearer to me to split the dependencies based on the `py-neurom` version.
However, I'll try to merge `py-pandas` so we'll have one line less.

GianlucaFicarelli committed:
Merge py-neurom dependencies

asanin-epfl reviewed (approved):
Looks good to me. As Matthias already approved it before, I'm approving.

tomdele reviewed (approved):
If CI is happy I am good

GianlucaFicarelli commented:
Thanks all, I'm going to merge! If anyone is still reviewing, please comment and I'll open a separate PR.

asanin-epfl commented:
Thank you for finishing this.
